packages/evm/src/eof/verify.ts

```diff
@@ -0,0 +1,519 @@
+ * A container cannot be marked both InitCode and DeploymentCode
+ * This flag is thus to distinguish between subcontainers, and also thus also allows for data section sizes
+ * lower than the size in the header in case of `InitCode`
+ */
+export enum ContainerSectionType {
+  InitCode, // Targeted by EOFCreate
+  DeploymentCode, // Targeted by RETURNCONTRACT
+  RuntimeCode, // "Default" runtime code
+}
+
+/**
+ * This method validates an EOF container deeply. It will validate the opcodes, validate the stack, and performs
+ * various checks such as checking for forbidden opcodes in certain modes, jumps to invalid places, etc.
+ * For more information, see "Code validation" of https://github.com/ipsilon/eof/blob/main/spec/eof.md
+ * This is a compilation of all the extra validation rules introduced by the various EIPs
+ * In particular, the stack validation EIP https://eips.ethereum.org/EIPS/eip-5450 is a big part here
+ * @param container EOFContainer to verify
+ * @param evm The EVM to run in (pulls opcodes from here)
+ * @param mode The validation mode to run in
+ * @returns Returns a Map which marks what ContainerSectionType each container is
+ * NOTE: this should likely not be a map, since a container section can only be of a single type, not multiple
+ */
+export function verifyCode(
+  container: EOFContainer,
+  evm: EVM,
+  mode: ContainerSectionType = ContainerSectionType.RuntimeCode
+) {
+  return validateOpcodes(container, evm, mode)
+}
+
+// Helper methods to read Int16s / Uint16s
+function readInt16(code: Uint8Array, start: number) {
+  return new DataView(code.buffer).getInt16(start)
+}
+
+function readUint16(code: Uint8Array, start: number) {
+  return new DataView(code.buffer).getUint16(start)
+}
+
+function validateOpcodes(
+  container: EOFContainer,
+  evm: EVM,
+  mode: ContainerSectionType = ContainerSectionType.RuntimeCode
+) {
+  // Track the intermediate bytes
+  const intermediateBytes = new Set<number>()
+  // Track the jump locations (for forward jumps it is unknown at the first pass if the byte is intermediate)
+  const jumpLocations = new Set<number>()
+
+  // Track the type of the container targets
+  // Should at the end of the analysis have all the containers
+  const containerTypeMap = new Map<number, ContainerSectionType>()
+
+  function addJump(location: number) {
+    if (intermediateBytes.has(location)) {
+      // When trying to JUMP into an intermediate byte: this is invalid
+      validationError(EOFError.InvalidRJUMP)
+    }
+    jumpLocations.add(location)
+  }
+
+  function addIntermediate(location: number) {
+    if (jumpLocations.has(location)) {
+      // When trying to add an intermediate to a location already JUMPed to: this is invalid
+      validationError(EOFError.InvalidRJUMP)
+    }
+    intermediateBytes.add(location)
+  }
+
+  // TODO (?) -> stackDelta currently only has active EOF opcodes, can use it directly (?)
+  // (so no need to generate the valid opcodeNumbers)
```

Comment:
Would second this thought if possible.  This feels like redoing the same thing everytime we are doing container validation.